### ISSUE TITLE
Checking whether there is data in the session, creates a session

### ISSUE THREAD

**Issue by disposable-ksa98** [symfony/symfony#6036]
UPDATE: This issue has 2 parts. The first one is explained in this post. The second one is explained in [this comment](https://github.com/symfony/symfony/issues/6036#issuecomment-10671945).

I noticed the website I am developing was sending a cookie in the response, even for anonymous requests (ie: not logged in, and deleted all cookies).

After trying everything and failing to find the problem, I created a "hello world" with a fresh install, and there was no cookie.

But as soon as I added this, the cookie came back: 

```
{% for flashMessage in app.session.flashbag.get('success') %}
    <div class="alert alert-success">
        <button type="button" class="close" data-dismiss="alert">×</button>
        {{ flashMessage }}
    </div>
{% endfor %}
```

So apparently `app.session.flashbag.get()` creates the session. I'm sure many other parts of my website are unintentionally creating a session, and that's why I couldn't find the culprit.

Is there a reason for this? Isn't it adding overhead that could have been avoided? I imagine this won't affect sites that are login-only, as they will need the session anyway for most of the visits. But sites that can be used anonymously might suffer a bit.

**Comment by netmikey** [comment]
We're starting to think about varnish caching as our site is growing. During our analysis, we stumbled upon the cacheability issue that @ureimers explained. Also, we're storing sessions into mysql and all this implicit session creation generates an awful lot of unnecessary load.

+1 for @Tobion's excellent behavior description and @ureimers quote about sane defaults btw. ;)

Also, github got me a bit confused with the merging and reverting... did the pull request eventually make it into Symfony @fabpot ? If so, what Symfony version would we need to migrate to, to be able to tame session creation?

**Comment by TerjeBr** [comment]
It is not in any symfony version yet.

I am still waiting for any response from @fabpot if we should go ahead with #6388 or not.

**Comment by weaverryan** [comment]
@Tobion I agree fully with your description as well - there's been some confusion between "is the session started" and "is there a session cookie - i.e. was the session started on a previous request". You explain that well. I think this is the right approach, so do you have time and knowledge of this area to look at #6388 to see if it's the right approach? We really either need to say "yes, this is the right approach let's finish this PR" or "no, it's not the right approach, let's work on a different PR".

Thanks :)

**Comment by weaverryan** [comment]
@TerjeBr And thanks for not letting this go - I know it's been a huge, long confusing road :)

**Comment by TerjeBr** [comment]
Thank you @weaverryan for giving your support for this issue. I absolutely agree with what you said. 

As I have said before, I am happy to continue work PR #6388 but I hesitate to do so before any one with some authority in this symfony2 community are willing to give me some positive feedback and say this is the way to go.

**Comment by TerjeBr** [comment]
@fabpot Any input on how to solve this? Should I update my #6388 pull request?

**Comment by ureimers** [comment]
@TerjeBr Thumbs up for not letting this one go for four years now!

**Comment by Koriit** [comment]
I fail to understand why this issue hasn't been resolved for 4 whole years now while it seems to me like an obvious bug!

There surely is a notice in Symfony documentation:
> While it is recommended to explicitly start a session, a session will actually start on demand, that is, if any session request is made to read/write session data.

While my reaction was "Well, sure, that's obvious thing to do" it never occured to me that reading data from non existent session(no session cookie) would first **create** empty session and only then try to check if the data I'm trying to get even exists. In my opinion this is wrong as this impacts performance and this is not an intuitive thing to do. When I read that notice, I thought that if read operation notices that there is no session to **load**(no session cookie) then it would return with NoData as there actually is no data to read. And I only stumbled on this thread by sheer luck on Google.

I think the problem here is that **to start a session** is actually ambiguous as it can mean either **creating completly new** session or **loading existing** session from the storage.

And as above, thumbs up for @TerjeBr.

**Comment by iloveyii** [comment]
I fixed the problem by entering the following in config.yml
```
session:
        cookie_domain: otp.softhem.se
```

**Comment by TerjeBr** [comment]
This is fixed by #24523 and can be closed.

**Comment by xabbuh** [comment]
#24523 isn't merged yet
